### PR TITLE
Implementation of space and time dependent bc.

### PR DESCRIPTION
**Pull request by iyer-arvind**
This implements the framework for a spatially and temporally
varying boundary condition. As an extension to the implementation
the sub-in-frv, sup-in-fa are now capable of taking in equations
in terms of x, y, z and t. Also, for the ease of usage, the
constants defined in the constants section are also available
in the equations.

### CODE REVIEW

File: pyfr/solvers/euler/kernels/bcs/sup-in-fa.mako [hunk 1]
```diff
@@ -1,11 +1,11 @@
 # -*- coding: utf-8 -*-
 <%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
 
-<%pyfr:macro name='bc_rsolve_state' params='ul, nl, ur'>
+<%pyfr:macro name='bc_rsolve_state' params='ul, nl, ur, ploc, t'>
     ur[0] = ${c['rho']};
 % for i in range(ndims):
```
Comment (FreddieWitherden) [inline pyfr/solvers/euler/kernels/bcs/sup-in-fa.mako]: `%for i, v in enumerate('uvw'[:ndims]):`

File: pyfr/solvers/euler/kernels/bcs/sup-in-fa.mako [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
-    ur[${i + 1}] = ${c['rho']*c['v'][i]};
+    ur[${i + 1}] = ${c['rho']}*${c['uvw'[i]]};
 % endfor
-    ur[${nvars - 1}] = ${c['p']/(c['gamma'] - 1) +\
-                         0.5*c['rho']*sum(v**2 for v in c['v'])};
+    ur[${nvars - 1}] = ${c['p']}/(${c['gamma']} - 1) +\
```
Comment (FreddieWitherden) [inline pyfr/solvers/euler/kernels/bcs/sup-in-fa.mako]: `${c['gamma'] - 1}` also avoids the need for `()`.  Also no need for the `\`.

File: pyfr/solvers/navstokes/kernels/bcs/sub-in-frv.mako [hunk 3]
```diff
@@ -2,14 +2,14 @@
 <%namespace module='pyfr.backends.base.makoutil' name='pyfr'/>
 <%include file='pyfr.solvers.navstokes.kernels.bcs.common'/>
 
-<%pyfr:macro name='bc_rsolve_state' params='ul, nl, ur'>
+<%pyfr:macro name='bc_rsolve_state' params='ul, nl, ur, ploc, t'>
     ur[0] = ${c['rho']};
 % for i in range(ndims):
-    ur[${i + 1}] = ${c['rho']*c['v'][i]};
+    ur[${i + 1}] = (${c['rho']}) * (${c['uvw'[i]]});
```
Comment (FreddieWitherden) [inline pyfr/solvers/navstokes/kernels/bcs/sub-in-frv.mako]: Space around `*`.

File: pyfr/solvers/navstokes/inters.py [hunk 4]
```diff
@@ -196,11 +200,11 @@ class NavierStokesSubInflowFrvBCInters(NavierStokesBaseBCInters):
...
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-
-        self._tpl_c['rho'], = self._eval_opts(['rho'])
-        self._tpl_c['v'] = self._eval_opts('uvw'[:self.ndims])
+    def __init__(self, be, lhs, elemap, cfgsect, cfg):
+        super().__init__(be, lhs, elemap, cfgsect, cfg)
+        tplc, self._ploc = self._exp_opts(
+            ['rho']+list('uvw'[:self.ndims]), lhs)
```
Comment (FreddieWitherden) [inline pyfr/solvers/navstokes/inters.py]: See if you can clean this up, `['rho', 'u', 'v', 'w'][:self.ndims + 1]`, maybe.

File: pyfr/solvers/baseadvec/inters.py [hunk 5]
```diff
@@ -94,3 +98,33 @@ def _eval_opts(self, opts, default=None):
             return [npeval(cfg.get(sect, k, default), cc) for k in opts]
         else:
             return [npeval(cfg.get(sect, k), cc) for k in opts]
+
+    def _exp_opts(self, opts, lhs, default=None):
```
Comment (FreddieWitherden) [inline pyfr/solvers/baseadvec/inters.py]: Ditch the `default` parameter if it is not used.

Comment (iyer-arvind) [inline pyfr/solvers/baseadvec/inters.py]: May be useful.
Say, if we decide u=0 v=0 w=0 by default for the subsonic inflow condition.
Not yet imeplemented, seeking view of others.

Or u=0 for no slip wall by default, but u,v,w
can be set to emulate a moving wall? Well this is usually a constant, but!

On Wed, May 13, 2015 at 12:14 PM, Freddie Witherden <
notifications@github.com> wrote:

> In pyfr/solvers/baseadvec/inters.py
> <https://github.com/vincentlab/PyFR/pull/52#discussion_r30221140>:
>
> > @@ -94,3 +98,33 @@ def _eval_opts(self, opts, default=None):
> >              return [npeval(cfg.get(sect, k, default), cc) for k in opts]
> >          else:
> >              return [npeval(cfg.get(sect, k), cc) for k in opts]
> > +
> > +    def _exp_opts(self, opts, lhs, default=None):
>
> Ditch the default parameter if it is not used.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vincentlab/PyFR/pull/52/files#r30221140>.
>



-- 
================================
Dr. Iyer Arvind Sundaram

The reasonable man adapts himself to the world; the unreasonable one
persists in trying to adapt the world to himself.

Therefore all progress depends on the unreasonable man

- George Bernard Shaw.

Comment (FreddieWitherden) [inline pyfr/solvers/baseadvec/inters.py]: Okay.  If you can do some local testing to ensure that everything works as expected when default is not None then I'm fine with it.  (I am always cautious about comitting code that might be untested.)

File: pyfr/solvers/euler/inters.py [hunk 6]
```diff
@@ -56,11 +56,11 @@ def __init__(self, *args, **kwargs):
...
     type = 'sup-in-fa'
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-
-        self._tpl_c['rho'], self._tpl_c['p'] = self._eval_opts(['rho', 'p'])
-        self._tpl_c['v'] = self._eval_opts('uvw'[:self.ndims])
+    def __init__(self, be, lhs, elemap, cfgsect, cfg):
+        super().__init__(be, lhs, elemap, cfgsect, cfg)
+        tplc, self._ploc = self._exp_opts(
```
Comment (FreddieWitherden) [inline pyfr/solvers/euler/inters.py]: Newline after super call.

File: pyfr/solvers/euler/inters.py [hunk 7]
```diff
@@ -56,11 +56,11 @@ def __init__(self, *args, **kwargs):
...
+            ['rho', 'p', 'u', 'v', 'w'][:self.ndims+2], lhs)
```
Comment (FreddieWitherden) [inline pyfr/solvers/euler/inters.py]: Space around binary `+`.  Put closing `)` on newline.

File: pyfr/solvers/navstokes/inters.py [hunk 8]
```diff
@@ -180,11 +184,11 @@ class NavierStokesSupInflowBCInters(NavierStokesBaseBCInters):
...
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-
-        self._tpl_c['rho'], self._tpl_c['p'] = self._eval_opts(['rho', 'p'])
-        self._tpl_c['v'] = self._eval_opts('uvw'[:self.ndims])
+    def __init__(self, be, lhs, elemap, cfgsect, cfg):
+        super().__init__(be, lhs, elemap, cfgsect, cfg)
+        tplc, self._ploc = self._exp_opts(
+            ['rho', 'p', 'u', 'v', 'w'][:self.ndims+2], lhs)
```
Comment (FreddieWitherden) [inline pyfr/solvers/navstokes/inters.py]: See above regarding formatting.

File: pyfr/solvers/navstokes/inters.py [hunk 9]
```diff
@@ -196,11 +200,11 @@ class NavierStokesSubInflowFrvBCInters(NavierStokesBaseBCInters):
...
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-
-        self._tpl_c['rho'], = self._eval_opts(['rho'])
-        self._tpl_c['v'] = self._eval_opts('uvw'[:self.ndims])
+    def __init__(self, be, lhs, elemap, cfgsect, cfg):
+        super().__init__(be, lhs, elemap, cfgsect, cfg)
+        tplc, self._ploc = self._exp_opts(
+            ['rho', 'u', 'v', 'w'][:self.ndims+1], lhs)
```
Comment (FreddieWitherden) [inline pyfr/solvers/navstokes/inters.py]: See above regarding formatting.

File: pyfr/solvers/baseadvec/inters.py (hunk 5 above)
Comment (iyer-arvind) [inline pyfr/solvers/baseadvec/inters.py]: I tested the default option settin the default u, v, w to zero when not
specified in sub-in-frv. I also modified the exp_option to be able to take
different defaults for different options. I am currently testing these,
will report back.

On Thu, May 14, 2015 at 10:48 PM, Freddie Witherden <
notifications@github.com> wrote:

> In pyfr/solvers/baseadvec/inters.py
> <https://github.com/vincentlab/PyFR/pull/52#discussion_r30366865>:
>
> > @@ -94,3 +98,33 @@ def _eval_opts(self, opts, default=None):
> >              return [npeval(cfg.get(sect, k, default), cc) for k in opts]
> >          else:
> >              return [npeval(cfg.get(sect, k), cc) for k in opts]
> > +
> > +    def _exp_opts(self, opts, lhs, default=None):
>
> Okay. If you can do some local testing to ensure that everything works as
> expected when default is not None then I'm fine with it. (I am always
> cautious about comitting code that might be untested.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vincentlab/PyFR/pull/52/files#r30366865>.
>



-- 
================================
Dr. Iyer Arvind Sundaram

The reasonable man adapts himself to the world; the unreasonable one
persists in trying to adapt the world to himself.

Therefore all progress depends on the unreasonable man

- George Bernard Shaw.

**Comment by FreddieWitherden** [comment]
If you can update the developer guide I am happy to merge.

**Comment by FreddieWitherden** [comment]
@pv101 If you're happy with the developer guide changes we can merge.

File: doc/src/user_guide.rst [hunk 10]
```diff
@@ -875,19 +875,27 @@ file with
 
         - ``rho`` --- density
 
-           *float*
+           *constant float or a function of physical location 'x', 'y', 'z'
```
Comment (pv101) [inline doc/src/user_guide.rst]: For consistency with elsewhere in the User Guide, I would just put *float* | *string* for the moment. Same comment applies in eight places below as well.

**Comment by pv101** [comment]
@FreddieWitherden just to confirm, do you mean User Guide? Or do we also need to update Developer Guide as well?

**Comment by FreddieWitherden** [comment]
Yes, the user guide, sorry.

**Pull request merged**